os/fs/smartfs/smartfs_utils.c

```diff
@@ -1028,35 +1063,98 @@ int smartfs_finddirentry(struct smartfs_mountpt_s *fs, struct smartfs_entry_s *d
  *
  ****************************************************************************/
 
-int smartfs_createentry(struct smartfs_mountpt_s *fs, uint16_t parentdirsector, const char *filename, uint16_t type, mode_t mode, struct smartfs_entry_s *direntry, uint16_t sectorno, FAR struct smartfs_ofile_s *sf)
+int smartfs_createentry(struct smartfs_mountpt_s *fs, uint16_t parentdirsector, struct smartfs_entry_s *ret_entry)
 {
+	fdbg("\n\n");
 	struct smart_read_write_s readwrite;
 	int ret;
-	uint16_t psector;
 	uint16_t nextsector;
-	uint16_t offset;
-	uint16_t found;
-	uint16_t entrysize;
-	struct smartfs_entry_header_s *entry;
 	struct smartfs_chain_header_s *chainheader;
 
 	/* Start at the 1st sector in the parent directory */
+	/* If we are in the context of smartfs_createentry, then smartfs_findentry did not find an available entry.
+	   Createentry has to chain a new nextsector to parentdirsector, and allocate the first entry.
+	 */
 
-	psector = parentdirsector;
-	found = FALSE;
-	entrysize = sizeof(struct smartfs_entry_header_s) + fs->fs_llformat.namesize;
+	readwrite.logsector = parentdirsector;
+	readwrite.count = fs->fs_llformat.availbytes;
+	readwrite.offset = 0;
+	readwrite.buffer = (uint8_t *)fs->fs_rwbuffer;
+	ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
+	if (ret < 0) {
+		goto errout;
+	}
 
-	/* Validate the name isn't too long */
+	chainheader = (struct smartfs_chain_header_s *)fs->fs_rwbuffer;
+	nextsector = SMARTFS_NEXTSECTOR(chainheader);
 
-	if (strlen(filename) > fs->fs_llformat.namesize) {
-		return -ENAMETOOLONG;
+	if (nextsector != SMARTFS_ERASEDSTATE_16BIT) {
+		fdbg("\nOops, something did not go as planned\n");
 	}
+	/* Allocate a new sector and chain it to the last one */
+
+	ret = FS_IOCTL(fs, BIOC_ALLOCSECT, 0xFFFF);
+	if (ret < 0) {
+		goto errout;
+	}
+
+	nextsector = (uint16_t)ret;
+
+	/* Chain the next sector into this sector sector */
```

Comment:
Ignore above comment, I think we'd better handle it in smartfs_open & sync_internal as I commented in smartfs_open, line 326

Reply:
Yes, I was a bit confused with this.
Hence I hadn't made any changes regarding this yet.

We can do as discussed for smartfs_open and smartfs_sync_internal.

We hold the new entry from being written till the sf buffer is full with data for the firstsector of the file and it is written to MTD.
So we write new entry and firstsector together to avoid chaining errors.